druid-shell/src/platform/x11/application.rs

```diff
@@ -404,4 +466,92 @@ impl Application {
+    // read it 16 bytes at a time just in case.
+    let mut read_buf = [0u8; 16];
+    loop {
+        match nix::unistd::read(idle_read, &mut read_buf[..]) {
+            Err(nix::Error::Sys(nix::errno::Errno::EINTR)) => {}
+            // According to write(2), this is the outcome of reading an empty, O_NONBLOCK
+            // pipe.
+            Err(nix::Error::Sys(nix::errno::Errno::EAGAIN)) => {
+                break;
+            }
+            Err(e) => {
+                return Err(e).context("Failed to read from idle pipe");
+            }
+            // According to write(2), this is the outcome of reading an O_NONBLOCK pipe
+            // when the other end has been closed. This shouldn't happen to us because we
+            // own both ends, but just in case.
+            Ok(0) => {
+                break;
+            }
+            Ok(_) => {}
+        }
+    }
+    Ok(())
+}
+
+/// Returns when there is an event ready to read from `conn`, or we got signalled by another thread
+/// writing into our idle pipe and the `timeout` has passed.
+// This was taken, with minor modifications, from the xclock_utc example in the x11rb crate.
+// https://raw.githubusercontent.com/psychon/x11rb/a6bd1453fd8e931394b9b1f2185fad48b7cca5fe/examples/xclock_utc.rs
+fn poll_with_timeout(conn: &Rc<XCBConnection>, idle: RawFd, timeout: Instant) -> Result<(), Error> {
+    use nix::poll::{poll, PollFd, PollFlags};
+    use std::os::raw::c_int;
+    use std::os::unix::io::AsRawFd;
+
+    let fd = conn.as_raw_fd();
+    let mut both_poll_fds = [
+        PollFd::new(fd, PollFlags::POLLIN),
+        PollFd::new(idle, PollFlags::POLLIN),
+    ];
+    let mut just_connection = [PollFd::new(fd, PollFlags::POLLIN)];
+    let mut poll_fds = &mut both_poll_fds[..];
+
+    // We start with no timeout in the poll call. If we get something from the idle handler, we'll
+    // start setting one.
+    let mut poll_timeout = -1;
+    loop {
+        let readable = |p: &PollFd| {
+            p.revents()
+                .unwrap_or_else(PollFlags::empty)
+                .contains(PollFlags::POLLIN)
+        };
+
+        match poll(poll_fds, poll_timeout) {
+            Ok(_) => {
+                if readable(&poll_fds[0]) {
+                    break;
+                }
+                if poll_fds.len() == 1 || readable(&poll_fds[1]) {
+                    // Now that we got signalled, stop polling from the idle pipe and use a timeout
+                    // instead.
+                    poll_fds = &mut just_connection;
+
+                    let now = Instant::now();
+                    if now >= timeout {
+                        break;
+                    } else {
+                        poll_timeout = c_int::try_from(timeout.duration_since(now).as_millis())
+                            .unwrap_or(c_int::max_value())
+                    }
+                }
+            }
+
+            Err(nix::Error::Sys(nix::errno::Errno::EINTR)) => {}
```

Comment:
What case does this handle?

Reply:
The C function `poll` returns `-EINTR` when a unix signal is received. This is necessary so that the signal can be handled (kernel-side reasons), but is quite useless for userspace. So, this case here just restarts the call.

After writing this: @jneem What about the `poll_timeout` argument? Some time could have passed before the signal came in and so `poll_timeout` might need to be recalculated.

Reply:
@psychon There's a good chance that I'm missing something obvious, but I think that once timeout gets set, then I recalculate it for every poll. Because if we ever set the timeout then we also set `poll_fds` to `just_connection`. Then every subsequent time through the loop the `poll_fds.len() == 1` will succeed and so we will either `break` or recalculate `poll_timeout`. Did I miss something?

Reply:
Ah, sorry for not looking carefully enough.

Reply:
Not at all! I'm very happy to have another pair of eyes on it!